File: tidal-dl/server/routers/jobs.py
"""任务批次管理 + JSON 导入 API"""
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form
from database import get_db_dependency
import json
import csv
import io

router = APIRouter(prefix="/api/jobs", tags=["Jobs"])


@router.get("")
def list_jobs(db=Depends(get_db_dependency)):
    """获取所有任务批次"""
    cursor = db.cursor()
    cursor.execute(
        "SELECT j.*, "
        "(SELECT COUNT(*) FROM tasks t WHERE t.job_id = j.id AND t.status = 'pending') as pending_count, "
        "(SELECT COUNT(*) FROM tasks t WHERE t.job_id = j.id AND t.status IN ('assigned','downloading','uploading')) as active_count "
        "FROM jobs j ORDER BY j.id DESC"
    )
    return cursor.fetchall()


@router.post("/import")
async def import_json(
    file: UploadFile = File(...),
    name: str = Form(""),
    quality: str = Form("LOSSLESS"),
    db=Depends(get_db_dependency)
):
    """导入 JSON 曲目列表创建任务批次"""
    content = await file.read()
    try:
        tracks = json.loads(content.decode("utf-8"))
    except (json.JSONDecodeError, UnicodeDecodeError) as e:
        raise HTTPException(status_code=400, detail=f"JSON 解析失败: {str(e)}")

    if not isinstance(tracks, list):
        raise HTTPException(status_code=400, detail="JSON 必须是数组格式")

    if not name:
        name = file.filename or "unnamed"

    cursor = db.cursor()

    # 创建 Job
    cursor.execute(
        "INSERT INTO jobs (name, source_file, total_tracks, target_quality, status) "
        "VALUES (%s, %s, %s, %s, 'running')",
        (name, file.filename, len(tracks), quality)
    )
    job_id = cursor.lastrowid

    # 批量创建 Tasks
    task_values = []
    for track in tracks:
        artist = track.get("artist", {})
        if isinstance(artist, dict):
            artist_name = artist.get("name", "Unknown")
        else:
            artist_name = str(artist)

        album = track.get("album", {})
        if isinstance(album, dict):
            album_title = album.get("title", "Unknown")
            album_id = album.get("id", 0)
        else:
            album_title = str(album)
            album_id = 0

        task_values.append((
            job_id,
            track.get("id", 0),
            track.get("title", "Unknown"),
            artist_name,
            album_title,
            album_id,
            track.get("trackNumber", 0),
            track.get("duration", 0),
            track.get("audioQuality", quality),
            track.get("isrc", ""),
        ))

    if task_values:
        cursor.executemany(
            "INSERT INTO tasks "
            "(job_id, track_id, title, artist, album, album_id, "
            "track_number, duration, audio_quality, isrc) "
            "VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)",
            task_values
        )

    db.commit()

    return {
        "message": f"导入成功: {len(tracks)} 首曲目",
        "job_id": job_id,
        "total_tracks": len(tracks)
    }


@router.post("/import-csv")
async def import_csv(
    file: UploadFile = File(...),
    name: str = Form(""),
    quality: str = Form("LOSSLESS"),
    db=Depends(get_db_dependency)
):
    """流式导入超大 CSV 列表创建任务批次"""
    if not name:
        name = file.filename or "unnamed_csv"

    cursor = db.cursor()
    # 先创建 Job 占位，total_tracks 稍后更新
    cursor.execute(
        "INSERT INTO jobs (name, source_file, total_tracks, target_quality, status) "
        "VALUES (%s, %s, %s, %s, 'running')",
        (name, file.filename, 0, quality)
    )
    job_id = cursor.lastrowid
    db.commit()

    total_count = 0
    batch_size = 5000
    task_values = []
    
    try:
        # 使用 TextIOWrapper 包装 spooled temporary file 进行流式读取
        text_file = io.TextIOWrapper(file.file, encoding='utf-8', errors='ignore')
        reader = csv.DictReader(text_file)
        
        for row in reader:
            track_id = row.get("id", "0").strip()
            if not track_id or not track_id.isdigit():
                continue

            track_id = int(track_id)
            isrc = row.get("isrc", "").strip()

            task_values.append((
                job_id,
                track_id,
                "Unknown",  # title
                "Unknown",  # artist
                "Unknown",  # album
                0,          # album_id
                0,          # track_number
                0,          # duration
                quality,
                isrc
            ))
            
            total_count += 1
            
            if len(task_values) >= batch_size:
                cursor.executemany(
                    "INSERT INTO tasks "
                    "(job_id, track_id, title, artist, album, album_id, "
                    "track_number, duration, audio_quality, isrc) "
                    "VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)",
                    task_values
                )
                db.commit()
                task_values = []
                
        # 插入剩余的
        if task_values:
            cursor.executemany(
                "INSERT INTO tasks "
                "(job_id, track_id, title, artist, album, album_id, "
                "track_number, duration, audio_quality, isrc) "
                "VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)",
                task_values
            )
            db.commit()
            
        # 更新 Job 的 total_tracks
        cursor.execute("UPDATE jobs SET total_tracks = %s WHERE id = %s", (total_count, job_id))
        db.commit()
        
    except Exception as e:
        cursor.execute("UPDATE jobs SET status = 'paused' WHERE id = %s", (job_id,))
        db.commit()
        raise HTTPException(status_code=400, detail=f"CSV 解析失败: {str(e)}")

    return {
        "message": f"流式导入成功: 写入 {total_count} 首曲目",
        "job_id": job_id,
        "total_tracks": total_count
    }


@router.get("/{job_id}")
def get_job(job_id: int, db=Depends(get_db_dependency)):
    """获取批次详情"""
    cursor = db.cursor()
    cursor.execute("SELECT * FROM jobs WHERE id = %s", (job_id,))
    job = cursor.fetchone()
    if not job:
        raise HTTPException(status_code=404, detail="批次不存在")

    # 获取状态统计
    cursor.execute(
        "SELECT status, COUNT(*) as count FROM tasks WHERE job_id = %s GROUP BY status",
        (job_id,)
    )
    status_counts = {row["status"]: row["count"] for row in cursor.fetchall()}
    job["status_counts"] = status_counts
    return job


@router.get("/{job_id}/tasks")
def get_job_tasks(
    job_id: int,
    status: str = "",
    page: int = 1,
    page_size: int = 50,
    db=Depends(get_db_dependency)
):
    """获取批次下的任务列表"""
    cursor = db.cursor()
    offset = (page - 1) * page_size

    if status:
        cursor.execute(
            "SELECT * FROM tasks WHERE job_id = %s AND status = %s "
            "ORDER BY id LIMIT %s OFFSET %s",
            (job_id, status, page_size, offset)
        )
    else:
        cursor.execute(
            "SELECT * FROM tasks WHERE job_id = %s ORDER BY id LIMIT %s OFFSET %s",
            (job_id, page_size, offset)
        )
    tasks = cursor.fetchall()

    # 总数
    if status:
        cursor.execute(
            "SELECT COUNT(*) as total FROM tasks WHERE job_id = %s AND status = %s",
            (job_id, status)
        )
    else:
        cursor.execute("SELECT COUNT(*) as total FROM tasks WHERE job_id = %s", (job_id,))
    total = cursor.fetchone()["total"]

    return {"tasks": tasks, "total": total, "page": page, "page_size": page_size}


@router.post("/{job_id}/retry-failed")
def retry_failed_tasks(job_id: int, db=Depends(get_db_dependency)):
    """重试批次中所有失败的任务"""
    cursor = db.cursor()
    cursor.execute(
        "UPDATE tasks SET status = 'pending', retry_count = 0, "
        "error_code = NULL, error_message = NULL, "
        "assigned_worker_id = NULL, assigned_account_id = NULL "
        "WHERE job_id = %s AND status IN ('failed', 'dead')",
        (job_id,)
    )
    affected = cursor.rowcount
    db.commit()
    return {"message": f"已重置 {affected} 个失败任务"}


@router.post("/{job_id}/pause")
def pause_job(job_id: int, db=Depends(get_db_dependency)):
    """暂停批次"""
    cursor = db.cursor()
    cursor.execute("UPDATE jobs SET status = 'paused' WHERE id = %s", (job_id,))
    db.commit()
    return {"message": "批次已暂停"}


@router.post("/{job_id}/resume")
def resume_job(job_id: int, db=Depends(get_db_dependency)):
    """恢复批次"""
    cursor = db.cursor()
    cursor.execute("UPDATE jobs SET status = 'running' WHERE id = %s", (job_id,))
    db.commit()
    return {"message": "批次已恢复"}

File: tidal-go/internal/server/handlers/jobs.go
package handlers

import (
	"encoding/csv"
	"encoding/json"
	"io"
	"net/http"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"

	"tidal-go/internal/models"
)

// ListJobs 批次列表。pending 从 jobs 字段推算(不扫大表)。
func (h *Handler) ListJobs(c *gin.Context) {
	var jobs []models.Job
	if err := h.DB.Select(&jobs, "SELECT * FROM jobs ORDER BY id DESC"); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	out := make([]gin.H, 0, len(jobs))
	for _, j := range jobs {
		pending := j.TotalTracks - j.Completed - j.Failed
		if pending < 0 {
			pending = 0
		}
		// 活跃数(主表小)
		var active int
		_ = h.DB.Get(&active,
			"SELECT COUNT(*) FROM tasks WHERE job_id=? AND status IN ('assigned','downloading','uploading')", j.ID)
		// 已下载总大小(从 export_groups 汇总,毫秒级)
		var totalBytes int64
		_ = h.DB.Get(&totalBytes, "SELECT COALESCE(SUM(total_size),0) FROM export_groups WHERE job_id=?", j.ID)
		out = append(out, gin.H{
			"id": j.ID, "name": j.Name, "source_file": j.SourceFile,
			"total_tracks": j.TotalTracks, "completed": j.Completed, "failed": j.Failed,
			"target_quality": j.TargetQuality, "status": j.Status,
			"created_at": j.CreatedAt, "updated_at": j.UpdatedAt,
			"pending_count": pending, "active_count": active, "total_bytes": totalBytes,
		})
	}
	c.JSON(http.StatusOK, out)
}

// GetJob 批次详情。
func (h *Handler) GetJob(c *gin.Context) {
	id := c.Param("id")
	var j models.Job
	if err := h.DB.Get(&j, "SELECT * FROM jobs WHERE id=?", id); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "批次不存在"})
		return
	}
	c.JSON(http.StatusOK, j)
}

// jsonTrack 导入 JSON 时的曲目结构(字段宽松)。
type jsonTrack struct {
	ID           int64           `json:"id"`
	Title        string          `json:"title"`
	Artist       json.RawMessage `json:"artist"`
	Album        json.RawMessage `json:"album"`
	TrackNumber  int             `json:"trackNumber"`
	Duration     int             `json:"duration"`
	AudioQuality string          `json:"audioQuality"`
	ISRC         string          `json:"isrc"`
}

// ImportJSON 导入 JSON 曲目列表创建批次。
func (h *Handler) ImportJSON(c *gin.Context) {
	name := c.PostForm("name")
	quality := c.DefaultPostForm("quality", "LOSSLESS")
	fh, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "缺少文件"})
		return
	}
	f, _ := fh.Open()
	defer f.Close()
	data, _ := io.ReadAll(f)

	var tracks []jsonTrack
	if err := json.Unmarshal(data, &tracks); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "JSON 解析失败: " + err.Error()})
		return
	}
	if name == "" {
		name = fh.Filename
	}

	res, err := h.DB.Exec(
		"INSERT INTO jobs (name, source_file, total_tracks, target_quality, status) VALUES (?,?,?,?,'running')",
		name, fh.Filename, len(tracks), quality)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	jobID, _ := res.LastInsertId()

	// 批量插入(每 5000 一批)
	h.bulkInsertTasks(jobID, quality, tracks)

	c.JSON(http.StatusOK, gin.H{"message": "导入成功", "job_id": jobID, "total_tracks": len(tracks)})
}

func (h *Handler) bulkInsertTasks(jobID int64, quality string, tracks []jsonTrack) {
	const batch = 5000
	buf := make([]jsonTrack, 0, batch)
	flush := func() {
		if len(buf) == 0 {
			return
		}
		var sb strings.Builder
		sb.WriteString("INSERT INTO tasks (job_id,track_id,title,artist,album,album_id,track_number,duration,audio_quality,isrc) VALUES ")
		args := make([]any, 0, len(buf)*10)
		for i, t := range buf {
			if i > 0 {
				sb.WriteByte(',')
			}
			sb.WriteString("(?,?,?,?,?,?,?,?,?,?)")
			artist := parseName(t.Artist)
			album, albumID := parseAlbum(t.Album)
			aq := t.AudioQuality
			if aq == "" {
				aq = quality
			}
			args = append(args, jobID, t.ID, defStr(t.Title, "Unknown"), artist, album, albumID,
				t.TrackNumber, t.Duration, aq, t.ISRC)
		}
		_, _ = h.DB.Exec(sb.String(), args...)
		buf = buf[:0]
	}
	for _, t := range tracks {
		buf = append(buf, t)
		if len(buf) >= batch {
			flush()
		}
	}
	flush()
}

// ImportCSV 流式导入超大 CSV(只需 id + isrc 列)。
func (h *Handler) ImportCSV(c *gin.Context) {
	name := c.PostForm("name")
	quality := c.DefaultPostForm("quality", "LOSSLESS")
	fh, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "缺少文件"})
		return
	}
	if name == "" {
		name = fh.Filename
	}

	res, _ := h.DB.Exec(
		"INSERT INTO jobs (name, source_file, total_tracks, target_quality, status) VALUES (?,?,0,?,'running')",
		name, fh.Filename, quality)
	jobID, _ := res.LastInsertId()

	f, _ := fh.Open()
	defer f.Close()
	reader := csv.NewReader(f)
	reader.FieldsPerRecord = -1

	header, err := reader.Read()
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "CSV 为空"})
		return
	}
	idCol, isrcCol := -1, -1
	for i, col := range header {
		switch strings.ToLower(strings.TrimSpace(col)) {
		case "id":
			idCol = i
		case "isrc":
			isrcCol = i
		}
	}
	if idCol < 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "CSV 缺少 id 列"})
		return
	}

	total := 0
	const batch = 5000
	var sb strings.Builder
	var args []any
	n := 0
	flush := func() {
		if n == 0 {
			return
		}
		_, _ = h.DB.Exec(sb.String(), args...)
		sb.Reset()
		args = args[:0]
		n = 0
	}
	begin := func() {
		sb.WriteString("INSERT INTO tasks (job_id,track_id,title,artist,album,album_id,track_number,duration,audio_quality,isrc) VALUES ")
	}
	begin()
	for {
		rec, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil || idCol >= len(rec) {
			continue
		}
		idStr := strings.TrimSpace(rec[idCol])
		tid, err := strconv.ParseInt(idStr, 10, 64)
		if err != nil {
			continue
		}
		isrc := ""
		if isrcCol >= 0 && isrcCol < len(rec) {
			isrc = strings.TrimSpace(rec[isrcCol])
		}
		if n > 0 {
			sb.WriteByte(',')
		}
		sb.WriteString("(?,?,?,?,?,?,?,?,?,?)")
		args = append(args, jobID, tid, "Unknown", "Unknown", "Unknown", 0, 0, 0, quality, isrc)
		n++
		total++
		if n >= batch {
			flush()
			begin()
		}
	}
	flush()

	_, _ = h.DB.Exec("UPDATE jobs SET total_tracks=? WHERE id=?", total, jobID)
	c.JSON(http.StatusOK, gin.H{"message": "导入成功", "job_id": jobID, "total_tracks": total})
}

// RetryFailed 把该批次的 dead 任务(在归档表)恢复为 pending(拉回活跃表)。
func (h *Handler) RetryFailed(c *gin.Context) {
	id := c.Param("id")
	// 从 archive 把 dead 拉回 tasks(status=pending),并从 archive 删除
	tx, err := h.DB.Beginx()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer tx.Rollback()

	res, _ := tx.Exec(
		"INSERT INTO tasks (id,job_id,track_id,title,artist,album,album_id,track_number,duration,audio_quality,isrc,status,max_retries) "+
			"SELECT id,job_id,track_id,title,artist,album,album_id,track_number,duration,audio_quality,isrc,'pending',3 "+
			"FROM tasks_archive WHERE job_id=? AND status='dead'", id)
	n, _ := res.RowsAffected()
	_, _ = tx.Exec("DELETE FROM tasks_archive WHERE job_id=? AND status='dead'", id)
	// 回退 jobs.failed 计数,批次重新 running
	_, _ = tx.Exec("UPDATE jobs SET failed=GREATEST(failed-?,0), status='running' WHERE id=?", n, id)
	if err := tx.Commit(); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "已重试失败任务", "count": n})
}

// PauseJob / ResumeJob 暂停/恢复批次(改 status,fetch 只取 running)。
func (h *Handler) PauseJob(c *gin.Context) {
	_, _ = h.DB.Exec("UPDATE jobs SET status='paused' WHERE id=? AND status='running'", c.Param("id"))
	c.JSON(http.StatusOK, gin.H{"message": "已暂停"})
}

func (h *Handler) ResumeJob(c *gin.Context) {
	_, _ = h.DB.Exec("UPDATE jobs SET status='running' WHERE id=? AND status='paused'", c.Param("id"))
	c.JSON(http.StatusOK, gin.H{"message": "已恢复"})
}

// GetJobTasks 批次任务列表(分页,活跃表 + 归档表联合)。
func (h *Handler) GetJobTasks(c *gin.Context) {
	id := c.Param("id")
	status := c.Query("status")
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "50"))
	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 50
	}
	offset := (page - 1) * pageSize

	// completed/dead 查归档表,其余查活跃表
	var tasks []models.Task
	var table string
	if status == "completed" || status == "dead" {
		table = "tasks_archive"
	} else {
		table = "tasks"
	}
	q := "SELECT * FROM " + table + " WHERE job_id=?"
	args := []any{id}
	if status != "" {
		q += " AND status=?"
		args = append(args, status)
	}
	q += " ORDER BY id LIMIT ? OFFSET ?"
	args = append(args, pageSize, offset)
	_ = h.DB.Select(&tasks, q, args...)
	c.JSON(http.StatusOK, gin.H{"tasks": tasks, "page": page, "page_size": pageSize})
}

// ---- helpers ----

func parseName(raw json.RawMessage) string {
	if len(raw) == 0 {
		return "Unknown"
	}
	var obj struct {
		Name string `json:"name"`
	}
	if json.Unmarshal(raw, &obj) == nil && obj.Name != "" {
		return obj.Name
	}
	var s string
	if json.Unmarshal(raw, &s) == nil && s != "" {
		return s
	}
	return "Unknown"
}

func parseAlbum(raw json.RawMessage) (string, int64) {
	if len(raw) == 0 {
		return "Unknown", 0
	}
	var obj struct {
		Title string `json:"title"`
		ID    int64  `json:"id"`
	}
	if json.Unmarshal(raw, &obj) == nil && obj.Title != "" {
		return obj.Title, obj.ID
	}
	var s string
	if json.Unmarshal(raw, &s) == nil && s != "" {
		return s, 0
	}
	return "Unknown", 0
}

func defStr(s, def string) string {
	if s == "" {
		return def
	}
	return s
}
